ddl/table.go

```diff
@@ -948,56 +956,188 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
 	return ver, t.UpdateTable(job.SchemaID, tblInfo)
 }
 
-// TODO: It may have the issue when two clients concurrently add partitions to a table.
-func onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+func checkAddPartition(t *meta.Meta, job *model.Job) (*model.TableInfo, *model.PartitionInfo, []*model.PartitionDefinition, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
 	partInfo := &model.PartitionInfo{}
-	err := job.DecodeArgs(&partInfo)
+	err = job.DecodeArgs(&partInfo)
 	if err != nil {
 		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+		return nil, nil, nil, errors.Trace(err)
+	}
+	partDefPointers := make([]*model.PartitionDefinition, 0, len(partInfo.Definitions))
+	for _, pd := range partInfo.Definitions {
+		partitionDefInMeta := tblInfo.FindPartitionDefinitionByName(pd.Name.L)
+		if partitionDefInMeta != nil {
+			if partitionDefInMeta.State == model.StatePublic {
```

Comment:
What if it is trying to rollback and one partition is in public?

Reply:
if one is public
1: you job is cancelled early in ddl_api check.
2: It has passed the early check and waited int the queue, in this interval, the same name partition has been added to public. So when this job get its turn to run, it's will failed here and cancelled directly. No rollback logic.